keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -1023,26 +1024,125 @@ this enhancement.
 
 #### Enforcement Actions
 
-For phase 1, all violations implicitly result in a `deny` enforcement action.
+For parity with admission webhooks, a validating policy may also emit audit
+annotations and warnings:
 
-For phase 2, we intend to support multiple enforcement actions.
+- [Audit
+  annotations](https://github.com/kubernetes/kubernetes/blob/97bbf07d3f3f20332912ee411fdf75ce84425e28/staging/src/k8s.io/api/admission/v1/types.go#L142)
+  are key/value pairs included in the audit event for an admission request. The
+  audit annotation key supplied by the policy definition author will be prefixed
+  with the name of the `ValidatingAdmissionPolicy` and policy binding, e.g.:
+  `mypolicy.mygroup.example.com/mybinding.mygroup.example.com/<annotation-key>`.
+- [Warnings](https://kubernetes.io/blog/2020/09/03/warnings/#admission-webhooks)
+  are string messages that are returned to API clients. Warning are returned for
+  both requests that are accepted and requests that are rejected.
 
-Use cases:
+`ValidatingAdmissionPolicy` may declare audit annotations in the policy
+definition. E.g.:
 
-- Cluster admin would like to rollout a policies, sometimes in bulk, without
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+...
+spec:
+  ...
+  validations:
+    - expression: <expression>
+  auditAnnotations:
+    - includeWhen: <expression> # optional field
+      key: "my-audit-key"
+      valueExpression: <expression that evaluates to a string>
+```
+
+Additionally, `ValidatingAdmissionPolicyBinding` resource may control how
+admission is enforced. This is performed using a single field. E.g.:
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicyBinding
+...
+spec:
+  enforcement: warn # optional field
+```
+
+- `deny`: Validation failures result in a denied request. (default beahvior if
+  field is unset)
+- `warn`: Validation failures are reported as warnings to the client.
+- `silent`: Validation failures are not reported to clients. Audit annotations
+          for any failed validations are still included in audit events (see
+          below).
+- (To disable audit annotations, delete the binding)
+
+Systems that need to aggregate validation failures may implement a [audit
+webhook
+backend](https://kubernetes.io/docs/tasks/debug/debug-cluster/audit/#webhook-backend). See
+below "Audit Events" for details.
+
+For singleton policies, the enforcement field will be set on the policy definition.
+
+Metrics will include enforcement so that cluster administrators can monitor the
```

Comment:
What kind of metrics are these? What metadata do they have associated with them (e.g. if Prometheus metrics does it include tags for the policy name or details about the violation? If so are there cardinality and/or security concerns?)

Reply:
Here's the metadata for the existing metric: https://github.com/kubernetes/kubernetes/blob/c63434aaff7bb3e2d89f89e72bb6c9ecc568a0db/staging/src/k8s.io/apiserver/pkg/admission/cel/metrics.go#L53

Reply:
Are there concerns about memory leaks due to metrics?

It looks like operators rely on https://github.com/kubernetes/kubernetes/pull/99385 to constraint unbounded labels?

Reply:
How many policies do you expect? That looks like it could get big.

@logicalhan is this cardinality too high?

Reply:
We reviewed cardinality here: https://github.com/kubernetes/kubernetes/pull/112994#discussion_r1001067238, and made a couple adjustments to manage it

The impact to cardinality for this change is that `enforcement_action` will also have `warn` and `audit` values (and increase from 2->4 possible values).